test/benchmark/__init__.py

```diff
@@ -0,0 +1,121 @@
+            self.timer = time.perf_counter()
+            return self
+
+        def end(self):
+            self.end_timer = time.perf_counter()
+
+        @property
+        def dif(self):
+            return self.end_timer - self.timer
+
+        def __exit__(self, exc_type, exc_val, exc_tb):
+            if not self.end_timer:
+                self.end_timer = time.perf_counter()
+            logger.info(f"{self.dif:.4f} seconds in {self.name}.")
+
+    class BenchmarkRunner:
+        gen_steps: typing.Tuple[str, ...] = (
+            "generate_early", "create_regions", "create_items", "set_rules", "generate_basic", "pre_fill")
+        rule_iterations: int = 100_000
+
+        @staticmethod
+        def format_times_from_counter(counter: collections.Counter[str], top: int = 5) -> str:
+            return "\n".join(f"  {time:.4f} in {name}" for name, time in counter.most_common(top))
+
+        def main(self):
+            for game in AutoWorld.AutoWorldRegister.world_types:
+                summary_data: typing.Dict[str, collections.Counter[str]] = {
+                    "empty_state": collections.Counter(),
+                    "all_state": collections.Counter(),
+                }
+                try:
+                    multiworld = MultiWorld(1)
+                    multiworld.game[1] = game
+                    multiworld.player_name = {1: "Tester"}
+                    multiworld.set_seed(0)
+                    multiworld.state = CollectionState(multiworld)
+                    args = Namespace()
+                    for name, option in AutoWorld.AutoWorldRegister.world_types[game].options_dataclass.type_hints.items():
+                        setattr(args, name, {
+                            1: option.from_any(getattr(option, "default"))
+                        })
+                    multiworld.set_options(args)
+
+                    gc.collect()
+                    for step in self.gen_steps:
+                        with TimeIt(f"{game} step {step}"):
+                            call_all(multiworld, step)
+                            gc.collect()
+
+                    all_state = multiworld.get_all_state(False)
+                    locations = multiworld.get_unfilled_locations()
+
+                    for location in locations:
+                        with TimeIt(f"{game} {self.rule_iterations} runs of {location}.access_rule(empty_state)") as t:
+                            for _ in range(self.rule_iterations):
+                                location.access_rule(multiworld.state)
+                            # if time is taken to disentangle complex ref chains,
+                            # this time should be attributed to the rule.
+                            gc.collect()
+                            t.end()
+                            summary_data["empty_state"][location.name] = t.dif
+
+                        with TimeIt(f"{game} {self.rule_iterations} runs of {location}.access_rule(all_state)"):
```

Comment:
The summary below is adding both times (empty and all state) because the old `t` is reused.
```suggestion
                        with TimeIt(f"{game} {self.rule_iterations} runs of {location}.access_rule(all_state)") as t:
```